CMakeLists.txt

```diff
@@ -18,16 +18,55 @@ if(MINGW)
 	endif()
 endif()
 
+#-----------------------------------------------------------------------------
+# Check which compiler to use
+
+if(${CMAKE_CXX_COMPILER_ID} MATCHES GNU)
+  INCLUDE(CheckCXXCompilerFlag)
+  CHECK_CXX_COMPILER_FLAG(-std=c++14 HAVE_STD14)
+  if(HAVE_STD14)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14 -pthread")
+    set(GCC_WARN_FLAGS "-Wnoexcept -Wstrict-null-sentinel -Wlogical-op")
+  else()
+    message(FATAL_ERROR "No advanced standard C++ support (-std=c++14 not defined).")
+  endif()
+elseif(${CMAKE_CXX_COMPILER_ID} MATCHES Clang)
+  set(CMAKE_COMPILER_IS_CLANGXX 1)
+  INCLUDE(CheckCXXCompilerFlag)
+  CHECK_CXX_COMPILER_FLAG(-std=c++14 HAVE_STD14)
+  if(HAVE_STD14)
+    if(APPLE)
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14 -stdlib=libc++")
+      set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} -stdlib=libc++")
+    else()
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
+    endif()
+  else()
+    message(FATAL_ERROR "No C++14 support for Clang version. Please upgrade Clang to a version supporting C++14.")
+  endif()
+endif()
+
 if (MSVC)
 	if(NOT MSVC14)
 		message(FATAL_ERROR "MSVC version not supported, try MSVC 14")
 	endif()
 endif()
 
-if (MINGW)
-	add_definitions(-D_WIN32_WINNT=0x0600)
+if(WIN32)
+  add_definitions(-DUNICODE)
+  add_definitions(-D_UNICODE)
+  add_definitions(-DWINVER=_WIN32_WINNT_WIN7)
+  add_definitions(-D_WIN32_WINNT=_WIN32_WINNT_WIN7)
```

Comment:
It is not working on MSVC with this line

5>c:\cygwin64\home\win2\galaxy42\src\c_tun_device.hpp(82): error C2039: 'stream_handle': is not a member of 'boost::asio::windows'
5>  C:\boost_1_61_0\boost/asio/windows/object_handle.hpp(28): note: see declaration of 'boost::asio::windows'
5>c:\cygwin64\home\win2\galaxy42\src\c_tun_device.hpp(82): error C2065: 'stream_handle': undeclared identifier
5>c:\cygwin64\home\win2\galaxy42\src\c_tun_device.hpp(82): error C2923: 'std::unique_ptr': 'stream_handle' is not a valid template type argument for parameter '_Ty'